jsonnet/kube-prometheus/prometheus-operator/prometheus-operator.libsonnet

```diff
@@ -0,0 +1,109 @@
+local krp = (import '../kube-rbac-proxy/container.libsonnet');
+local prometheusOperator = import 'github.com/prometheus-operator/prometheus-operator/jsonnet/prometheus-operator/prometheus-operator.libsonnet';
+
+local defaults = {
+  local defaults = self,
+  name: 'prometheus-operator',
+  namespace: error 'must provide namespace',
+  version: error 'must provide version',
+  image: error 'must provide image',
+  configReloaderImage: error 'must provide config reloader image',
+  resources: {
+    limits: { cpu: '200m', memory: '200Mi' },
+    requests: { cpu: '100m', memory: '100Mi' },
+  },
+  commonLabels:: {
+    'app.kubernetes.io/name': defaults.name,
+    'app.kubernetes.io/version': defaults.version,
+    'app.kubernetes.io/component': 'controller',
+    'app.kubernetes.io/part-of': 'kube-prometheus',
+  },
+  selectorLabels:: {
+    [labelName]: defaults.commonLabels[labelName]
+    for labelName in std.objectFields(defaults.commonLabels)
+    if !std.setMember(labelName, ['app.kubernetes.io/version'])
+  },
+};
+
+function(params) {
+  local po = self,
+  config:: defaults + params,
+  // Safety check
+  assert std.isObject(po.config.resources),
+
+  //TODO(paulfantom): it would be better to include it on the same level as self.
+  local polib = prometheusOperator(po.config),
+
+  '0alertmanagerConfigCustomResourceDefinition': polib['0alertmanagerConfigCustomResourceDefinition'],
+  '0alertmanagerCustomResourceDefinition': polib['0alertmanagerCustomResourceDefinition'],
+  '0podmonitorCustomResourceDefinition': polib['0podmonitorCustomResourceDefinition'],
+  '0probeCustomResourceDefinition': polib['0probeCustomResourceDefinition'],
+  '0prometheusCustomResourceDefinition': polib['0prometheusCustomResourceDefinition'],
+  '0prometheusruleCustomResourceDefinition': polib['0prometheusruleCustomResourceDefinition'],
+  '0servicemonitorCustomResourceDefinition': polib['0servicemonitorCustomResourceDefinition'],
+  '0thanosrulerCustomResourceDefinition': polib['0thanosrulerCustomResourceDefinition'],
+
+  serviceAccount: polib.serviceAccount,
```

Comment:
I don't have any other idea how to include `prometheus-operator.libsonnet` without those mappings. The only other idea requires another refactor in an upstream library to make it look similar to https://github.com/kubernetes/kube-state-metrics/blob/master/jsonnet/kube-state-metrics/kube-state-metrics.libsonnet and then include it the same way as we do it in https://github.com/prometheus-operator/kube-prometheus/blob/master/jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet#L28.

Either way, the current version still allows end-users to customize prometheus-operator while using kube-prometheus. This means a potential refactor can happen later.

Reply:
Why don't we just create a prometheus instance from the config and then override it later on?
Something like this (not tested)
```jsonnet
function(params) {
  local config = defaults + params,
  // Safety check
  assert std.isObject(config.resources),

  //TODO(paulfantom): it would be better to include it on the same level as self.
  prometheusOperator(config) {
  local po = self,
  config:: config,
  // Do the overrides
},
```

It would be easier to reason about and you won't have to change if you add anything on upstream.

Reply:
@paulfantom Yeah, I just yolo scribbled it. (I just fixed the example now) I can check out it locally and play with when I've time. I remember facing a similar issue, I can also try to find out a working example.

Reply:
Nice! I didn't try doing it this way! It looks better and it works, thanks! :100: